components/service/src/service_manager.rs

```diff
@@ -0,0 +1,80 @@
+use std::sync::{atomic::Ordering, Arc};
+
+use atomic::Atomic;
+use crossbeam::channel::SendError;
+use tikv_util::mpsc;
+
+use crate::service_event::ServiceEvent;
+
+#[repr(u8)]
+#[derive(Debug, Copy, Clone, PartialEq)]
+enum GrpcServiceStatus {
+    Init,
+    Serving,
+    NotServing,
+}
+
+#[derive(Clone)]
+pub struct GrpcServiceManager {
+    status: Arc<Atomic<GrpcServiceStatus>>,
+    service_router: mpsc::Sender<ServiceEvent>,
+}
+
+impl GrpcServiceManager {
+    fn build(router: mpsc::Sender<ServiceEvent>, status: GrpcServiceStatus) -> Self {
+        Self {
+            status: Arc::new(Atomic::new(status)),
+            service_router: router,
+        }
+    }
+
+    /// Generate a formal GrpcServiceManager.
+    pub fn new(router: mpsc::Sender<ServiceEvent>) -> Self {
+        Self::build(router, GrpcServiceStatus::Serving)
+    }
+
+    /// Only for test.
+    /// Generate a dummy GrpcServiceManager.
+    pub fn dummy() -> Self {
+        let (router, _) = mpsc::unbounded();
+        Self::build(router, GrpcServiceStatus::Init)
+    }
+
+    /// Send message to outer handler to notify PAUSE grpc server.
+    pub fn pause(&mut self) -> Result<(), SendError<ServiceEvent>> {
+        if self.is_paused() {
+            // Already in PAUSE.
+            return Ok(());
+        }
+        let result = self.service_router.send(ServiceEvent::PauseGrpc);
+        if result.is_ok() {
+            self.status
+                .store(GrpcServiceStatus::NotServing, Ordering::Relaxed);
```

Comment:
this status is really not very useful, as the real pause/resume status is not known at this layer at all. I think this can be removed and the check should be added where the actual pause/resume is done.